File: frontend/src/types/document-improvement.ts
export type DocumentImprovementType =
  | "CLARITY"
  | "CONCISENESS"
  | "PROFESSIONAL_TONE"
  | "JOB_ALIGNMENT"
  | "COMPANY_ALIGNMENT"
  | "SKILL_EMPHASIS"
  | "EXPERIENCE_EMPHASIS"
  | "PROJECT_EMPHASIS"
  | "ACHIEVEMENT_EMPHASIS"
  | "STRUCTURE"
  | "GRAMMAR"
  | "LENGTH_REDUCTION"
  | "LENGTH_EXPANSION"
  | "CUSTOM";

export type DocumentImprovementRunStatus =
  | "PENDING"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED"
  | "INVALID_OUTPUT"
  | "REVIEW_REQUIRED"
  | "APPLIED"
  | "REJECTED";

export type DocumentImprovementSuggestionStatus = "PENDING" | "APPROVED" | "REJECTED" | "APPLIED";
export type DocumentImprovementRiskLevel = "LOW" | "MEDIUM" | "HIGH";
export type DocumentImprovementChangeType = "REWRITE" | "ADD" | "DELETE" | "STRUCTURE" | "GRAMMAR" | "TONE" | "LENGTH";

export type DocumentImprovementCreatePayload = {
  improvement_type: DocumentImprovementType;
  custom_instruction?: string | null;
  base_version_id?: number | null;
  target_min_length?: number | null;
  target_max_length?: number | null;
  target_tone?: string | null;
  source_ids?: number[];
};

export type DocumentImprovementSuggestionPublic = {
  id: number;
  run_id: number;
  paragraph_index: number;
  sentence_index: number;
  original_text: string;
  suggested_text: string;
  change_type: DocumentImprovementChangeType;
  reason: string;
  evidence: Record<string, unknown>[];
  risk_level: DocumentImprovementRiskLevel;
  status: DocumentImprovementSuggestionStatus;
  selected: boolean;
  applied_at: string | null;
  created_at: string;
  updated_at: string;
};

export type DocumentImprovementRunPublic = {
  id: number;
  user_id: number;
  application_document_id: number;
  base_version_id: number;
  result_version_id: number | null;
  status: DocumentImprovementRunStatus;
  improvement_type: DocumentImprovementType;
  custom_instruction: string | null;
  target_min_length: number | null;
  target_max_length: number | null;
  target_tone: string | null;
  provider: string;
  model: string | null;
  prompt_version: string;
  schema_version: string;
  input_hash: string;
  source_hash: string;
  outdated: boolean;
  started_at: string;
  completed_at: string | null;
  error_code: string | null;
  safe_error_message: string | null;
  overall_feedback: string | null;
  suggested_title: string | null;
  suggested_content: string | null;
  warnings: string[];
  missing_information: string[];
  confidence: Record<string, number>;
  usage_metadata: Record<string, unknown>;
  created_at: string;
  updated_at: string;
  suggestions: DocumentImprovementSuggestionPublic[];
  sources: Array<Record<string, unknown>>;
  actions: Array<Record<string, unknown>>;
};

export type DocumentImprovementListData = {
  items: DocumentImprovementRunPublic[];
  page: number;
  size: number;
  total: number;
  total_pages: number;
};

export type DocumentImprovementApplyData = {
  applied: boolean;
  version_id: number;
  version_number: number;
  applied_suggestion_count: number;
};

File: backend/app/schemas/document_improvement.py
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, model_validator

from app.models.document_improvement import (
    DocumentImprovementActionType,
    DocumentImprovementChangeType,
    DocumentImprovementRiskLevel,
    DocumentImprovementRunStatus,
    DocumentImprovementSourceType,
    DocumentImprovementSuggestionStatus,
    DocumentImprovementType,
)


class DocumentImprovementEvidence(BaseModel):
    source_type: DocumentImprovementSourceType
    source_id: str
    source_text: str = Field(min_length=1, max_length=1000)


class DocumentImprovementSentenceSuggestionData(BaseModel):
    id: str | None = None
    paragraph_index: int = Field(default=0, ge=0)
    sentence_index: int = Field(default=0, ge=0)
    original_text: str = Field(min_length=1)
    suggested_text: str = Field(min_length=1)
    change_type: DocumentImprovementChangeType
    reason: str = Field(min_length=1)
    evidence: list[DocumentImprovementEvidence] = Field(default_factory=list)
    risk_level: DocumentImprovementRiskLevel = DocumentImprovementRiskLevel.LOW
    selected: bool = True


class DocumentImprovementStructuredData(BaseModel):
    summary: str
    overall_feedback: str
    suggested_title: str | None = None
    suggested_content: str
    sentence_suggestions: list[DocumentImprovementSentenceSuggestionData]
    warnings: list[str] = Field(default_factory=list)
    missing_information: list[str] = Field(default_factory=list)
    used_sources: list[DocumentImprovementEvidence] = Field(default_factory=list)
    confidence: dict[str, float] = Field(default_factory=dict)


class DocumentImprovementCreateRequest(BaseModel):
    improvement_type: DocumentImprovementType
    custom_instruction: str | None = Field(default=None, max_length=2000)
    base_version_id: int | None = None
    target_min_length: int | None = Field(default=None, ge=1, le=10000)
    target_max_length: int | None = Field(default=None, ge=1, le=10000)
    target_tone: str | None = Field(default=None, max_length=80)
    source_ids: list[int] = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_lengths(self):
        if self.target_min_length and self.target_max_length and self.target_min_length > self.target_max_length:
            raise ValueError("target_min_length must be less than or equal to target_max_length")
        if self.improvement_type == DocumentImprovementType.CUSTOM and not self.custom_instruction:
            raise ValueError("custom_instruction is required for CUSTOM improvement")
        return self


class DocumentImprovementSuggestionUpdate(BaseModel):
    status: DocumentImprovementSuggestionStatus | None = None
    selected: bool | None = None


class DocumentImprovementApplyRequest(BaseModel):
    apply_all: bool = False
    suggestion_ids: list[int] = Field(default_factory=list)
    version_title: str | None = Field(default=None, max_length=200)


class DocumentImprovementSourcePublic(BaseModel):
    id: int
    run_id: int
    source_type: DocumentImprovementSourceType
    source_id: str
    source_hash: str
    source_snapshot: dict[str, Any]
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentImprovementSuggestionPublic(BaseModel):
    id: int
    run_id: int
    paragraph_index: int
    sentence_index: int
    original_text: str
    suggested_text: str
    change_type: DocumentImprovementChangeType
    reason: str
    evidence: list[dict[str, Any]]
    risk_level: DocumentImprovementRiskLevel
    status: DocumentImprovementSuggestionStatus
    selected: bool
    applied_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class DocumentImprovementActionPublic(BaseModel):
    id: int
    user_id: int
    run_id: int
    suggestion_id: int | None
    action: DocumentImprovementActionType
    previous_text: str | None
    new_text: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class DocumentImprovementRunPublic(BaseModel):
    id: int
    user_id: int
    application_document_id: int
    base_version_id: int
    result_version_id: int | None
    status: DocumentImprovementRunStatus
    improvement_type: DocumentImprovementType
    custom_instruction: str | None
    target_min_length: int | None
    target_max_length: int | None
    target_tone: str | None
    provider: str
    model: str | None
    prompt_version: str
    schema_version: str
    input_hash: str
    source_hash: str
    outdated: bool
    started_at: datetime
    completed_at: datetime | None
    error_code: str | None
    safe_error_message: str | None
    overall_feedback: str | None
    suggested_title: str | None
    suggested_content: str | None
    warnings: list[str] = Field(default_factory=list)
    missing_information: list[str] = Field(default_factory=list)
    confidence: dict[str, float] = Field(default_factory=dict)
    usage_metadata: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime
    suggestions: list[DocumentImprovementSuggestionPublic] = Field(default_factory=list)
    sources: list[DocumentImprovementSourcePublic] = Field(default_factory=list)
    actions: list[DocumentImprovementActionPublic] = Field(default_factory=list)

    model_config = {"from_attributes": True}


class DocumentImprovementListData(BaseModel):
    items: list[DocumentImprovementRunPublic]
    page: int
    size: int
    total: int
    total_pages: int


class DocumentImprovementApplyData(BaseModel):
    applied: bool
    version_id: int
    version_number: int
    applied_suggestion_count: int


class DocumentImprovementRejectData(BaseModel):
    rejected: Literal[True] = True


class DocumentImprovementDeletedData(BaseModel):
    deleted: bool
